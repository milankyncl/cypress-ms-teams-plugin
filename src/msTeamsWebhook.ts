import CypressCommandLine from 'cypress';
import { IncomingWebhook } from "ms-teams-webhook";
import { Options } from ".";

// Should implement CypressCommandLine.CypressRunResult
interface TestRun {
    tests: {
        state: string;
        title: string[];
    }[];
    stats: {
        failures: number;
    };
    spec: {
        name: string;
    };
}

interface RunResult {
    totalDuration: number;
    totalFailed: number;
    runs: TestRun[];
}

export function getResultMessagePayload(options: Omit<Options, 'msTeamsWebhookUrl'>, result: RunResult) {
    const failedCases = result.runs.map((run) => {
        if (run.stats.failures > 0) {
            return run.tests.map((test, i) => test.state === 'failed'
                ? `${i + 1}) ${test.title.join(' / ')} (${run.spec.name})` : null).filter((test): test is string => test !== null);
        }
        return [];
    }).flat();
    const failed = result.totalFailed > 0;

    return {
        '@type': 'MessageCard',
        '@context': 'https://schema.org/extensions',
        summary: 'Cypress test results',
        themeColor: failed ? '#FF0000' : '#009a00',
        title: failed ? `${result.totalFailed} test case(s) failed` : 'All test cases passed',
        sections: ([
            options.ref ? {
                activityTitle: options.ref,
                activitySubtitle: `Test run completed finished in ${(result.totalDuration / 1000).toString()}s`,
            } : undefined,
            failed ? {
                text: failed ? ['<strong>Failed test cases:</strong>', ...failedCases].join('<br />') : undefined,
            } : undefined
        ]).filter((section) => section !== undefined),
        potentialAction: options.runUrl ? [
            {
                '@context': 'http://schema.org',
                '@type': 'ViewAction',
                name: 'Open Test Run Report',
                target: [options.runUrl],
            },
        ] : undefined,
    }
}

export async function sendFinishedResults(
    options: Options,
    testResults: CypressCommandLine.CypressRunResult,
) {
    const webhook = new IncomingWebhook(options.msTeamsWebhookUrl);
    await webhook.send(JSON.stringify(getResultMessagePayload(
        options,
        testResults,
    )));
}