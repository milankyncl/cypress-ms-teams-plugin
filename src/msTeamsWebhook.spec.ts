import { describe, expect, it } from '@jest/globals';
import { getResultMessagePayload } from "./msTeamsWebhook";

describe('MS Teams webhook', () => {
	describe('#getResultMessagePayload', () => {
		describe('with one test case failed', () => {
			it('should create failure message payload', () => {
				expect(
					getResultMessagePayload({
						ref: 'v0.0.1',
					}, {
						totalFailed: 1,
						totalDuration: 2_000,
						runs: [
							{
								stats: {
									failures: 1,
								},
								tests: [
									{
										title: ['Test suite #1', 'do this and that'],
										state: 'failed',
									},
								],
								spec: {
									name: 'file1.spec.js',
								},
							}
						],
					}),
				).toMatchSnapshot();
			});

			it('should create success message payload', () => {
				expect(
					getResultMessagePayload({
						ref: 'v0.0.2',
					}, {
						totalFailed: 0,
						totalDuration: 5_000,
						runs: [
							{
								stats: {
									failures: 0,
								},
								tests: [
									{
										title: ['Test suite #2', 'should do this and this'],
										state: 'passed',
									},
								],
								spec: {
									name: 'file2.spec.js',
								},
							}
						],
					}),
				).toMatchSnapshot();
			});
		});
	});
});