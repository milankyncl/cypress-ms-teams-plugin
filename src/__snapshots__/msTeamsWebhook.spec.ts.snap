// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MS Teams webhook #getResultMessagePayload with one test case failed should create failure message payload 1`] = `
{
  "@context": "https://schema.org/extensions",
  "@type": "MessageCard",
  "potentialAction": undefined,
  "sections": [
    {
      "activitySubtitle": "Test run completed finished in 2s",
      "activityTitle": "v0.0.1",
    },
    {
      "text": "<strong>Failed test cases:</strong><br />1) Test suite #1 / do this and that (file1.spec.js)",
    },
  ],
  "summary": "Cypress test results",
  "themeColor": "#FF0000",
  "title": "1 test case(s) failed",
}
`;

exports[`MS Teams webhook #getResultMessagePayload with one test case failed should create success message payload 1`] = `
{
  "@context": "https://schema.org/extensions",
  "@type": "MessageCard",
  "potentialAction": undefined,
  "sections": [
    {
      "activitySubtitle": "Test run completed finished in 5s",
      "activityTitle": "v0.0.2",
    },
  ],
  "summary": "Cypress test results",
  "themeColor": "#009a00",
  "title": "All test cases passed",
}
`;
